src/next_array.rs

```diff
@@ -0,0 +1,144 @@
+use core::mem::{self, MaybeUninit};
+use core::ptr;
+
+/// An array of at most `N` elements.
+struct ArrayBuilder<T, const N: usize> {
+    /// The (possibly uninitialized) elements of the `ArrayBuilder`.
+    ///
+    /// # Safety
+    ///
+    /// The elements of `arr[..len]` are valid `T`s.
+    arr: [MaybeUninit<T>; N],
+
+    /// The number of leading elements of `arr` that are valid `T`s, len <= N.
+    len: usize,
+}
+
+impl<T, const N: usize> ArrayBuilder<T, N> {
+    /// Initializes a new, empty `ArrayBuilder`.
+    pub fn new() -> Self {
+        // SAFETY: The safety invariant of `arr` trivially holds for `len = 0`.
+        Self {
+            arr: [(); N].map(|_| MaybeUninit::uninit()),
+            len: 0,
+        }
+    }
+
+    /// Pushes `value` onto the end of the array.
+    ///
+    /// # Panics
+    ///
+    /// This panics if `self.len >= N` or if `self.len == usize::MAX`.
+    pub fn push(&mut self, value: T) {
+        // PANICS: This will panic if `self.len >= N`.
+        // SAFETY: The safety invariant of `self.arr` applies to elements at
+        // indices `0..self.len` — not to the element at `self.len`. Writing to
+        // the element at index `self.len` therefore does not violate the safety
+        // invariant of `self.arr`. Even if this line panics, we have not
+        // created any intermediate invalid state.
+        self.arr[self.len] = MaybeUninit::new(value);
+        // PANICS: This will panic if `self.len == usize::MAX`.
+        // SAFETY: By invariant on `self.arr`, all elements at indicies
+        // `0..self.len` are valid. Due to the above write, the element at
+        // `self.len` is now also valid. Consequently, all elements at indicies
+        // `0..(self.len + 1)` are valid, and `self.len` can be safely
+        // incremented without violating `self.arr`'s invariant. It is fine if
+        // this increment panics, as we have not created any intermediate
+        // invalid state.
+        self.len = match self.len.checked_add(1) {
+            Some(sum) => sum,
+            None => panic!("`self.len == usize::MAX`; cannot increment `len`"),
+        };
+    }
+
+    /// Consumes the elements in the `ArrayBuilder` and returns them as an array
+    /// `[T; N]`.
+    ///
+    /// If `self.len() < N`, this returns `None`.
+    pub fn take(&mut self) -> Option<[T; N]> {
+        if self.len == N {
+            // SAFETY: Decreasing the value of `self.len` cannot violate the
+            // safety invariant on `self.arr`.
+            self.len = 0;
+
+            // SAFETY: Since `self.len` is 0, `self.arr` may safely contain
+            // uninitialized elements.
+            let arr = mem::replace(&mut self.arr, [(); N].map(|_| MaybeUninit::uninit()));
```

Comment:
could be simplified replace all of `self` with `ArrayBuilder::new()`?

Reply:
I don't believe so. I agree, it would be ideal if we could simply write:
```rust
let Self { arr, len } = mem::replace(self, Self::new());
```
instead. However, because `ArrayBuilder` has a non-trivial `Drop` implementation, we cannot move-destructure it in that manner. Instead, we need to read and write each field individually.

Reply:
I was going to suggest using `std::ptr::read` instead of `std::mem::replace`ing with `[uninit(); N]`, but in release mode they seem to be optimized to the same thing (I assume the optimizer knows that writing `uninit()` is a no-op), so whichever is more readable is probably better. ([godbolt link](https://rust.godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywYgArAA5SjgBk8BkwAOXcAI0xiEABmPwAHVAVCOwYXNw9vROTUgSCQ8JYomPjLTGtbASECJmICDPdPX3LKtJq6ggKwyOi4v3NOxqyWwfruopL%2BgEpLVBNiZHYOM0wAanN0EBA2Fm2AWSYATyi5BmDCAFJYgCFLjQBBe4eEkwiNglMbNYfiYmOAGqYZDXJwAFVIazQDHMa1CIDWZjwAC9MNdsGtLgB2O6PNb4tboJi1BGXLw3Q4nTBnC4EUFg9HXG5wskAEVIzwJa3onkRKVRHMe2NZz2eeBYCVo9Mh0Nh8L5KLRsQxv3%2BRyBINi4MhoXRmJxnIJr3esoIayoDDWIQA7hBpmsALR6oQVKj63EPLlcl20N3Yj1ewOE4lMUnk03utaU07nc4EbYmWOEO36kW3FledmGoMEnkIjSCz1B4XZ1OivFGt7my2vBQICCXABMADYWCYzVYqJCAG5iEzsNZg%2B3%2B0tczsAOiJtTJNwnPLZmNirKjxxjtO2NogvbcmGmTNHBLnjExjbuS7WXH3FfxJaF17Wxura1qAGtMAB9LgNlttjuu%2B1OsqawAPIJFUDCgjODLprqmZ6iO95cngbpHpaYBgNcy6hO6B6BsQmAEAslqhAISoBsWWIiohh6uuOPKLsuGhXkWgb0GadRqgxiIwkwVDrAhLE5h8WwgGBMQgPhTDoBAmzbGJ2xUMQqAsO%2B%2BFUN%2BzYTlOTDTOOojmHae7UTelHMTmQjKZgEAcccRmCbeTz3o%2BFrPkwb7vo2Gm/hs/6OnqoHgZB5LQcysFphiAk5shPm%2BnRx7oZhcI4cZXr4YRxDEaRZkUVRgljrR9GJUxty4dyBFrDZRxcbJOyYHsEmYJKTBLF57YxVQk4hpCM52kyGasuOLBMAkDZYk477Yk4K5UjScYJkmBCGXuJUpRZbDWX8xyDcNo1OBUdWTdxCi8fxOJrPtLDjkwCgKO4H60imwrTHZXIOW9QqNrEM6rOgbLPIMtjID8CIWq1ZqqoCwKgiYsSNpCsPogB/lgWkQU3DDjZ9QjcFAYlEPqsC2yvh%2Bl4rY5Tyfd9Si/XBjwA3gQM3CDDBgz8m0E5qTgY/DmPKkjQEBajWrfQj6bY%2BFXH4xqRNuR%2BvM3Bwsy0JwXi8J43C8KgnAAEpmB28yLPxn08KQ8YcFoz2kC%2B3iNuOGjNi2zaxFwPheLEWKuwAnPonCSGrmia5wvAKCABZmxbpBwLASBoBKdDROQlCxwk8cxOcwAIAQtBHHwdAENEIcQBEAekBEwR1EcnAm2XzDEEcwERNowJmybsdsIIwEMNnJdYBEJjAE4Yi0CHGukFgQ1GOI5u8Pg%2BE2Hg3aYCPWj%2BKowLtssJvBPnyvT/oeARP8dcuFgJefOKVe8IvxARMkmCsnVhjALQwSgNPsxUAYwAKACeCYNaoFGCXxkIIEQYh2BSBAfIJQagS66C4PoJ%2BEk9b7wiCHSAsxUAowECPB0LgwIOnoIvKUS5Yia2vsQPAWB0FQGYGwEAmB8DgVINufsnBGwaEbDwS2Vhm5pAcAwZwrgmh6ECMEHoxQ%2BgIKSCkcCwxPDSNyOBCYvQYgIN4fPaotR6jyL0Bo8CHRxjiMmFIyw2iGjCKyOo8xKjJFqNmAoA2SwJBKxVv7PeWsOBrFMOYIG6dM7ZzWBAXAhASAnmdtMXg4dFaRyQPMAgrwLAUAgMnVOoRWDLB8QQIGXBxweztrwRhoTKFbAQfwUBohxCQPKdAlQ6g97wNINaf4CRL6uI4KrUg6sV6eOAu2RJaxUBuiyX4vAGcs5VQgC4OO9BiDhK4JE02AdLbW0kFiccnt3axFiM2HwUgHZSAQbvP2pA9hcA0AWbpgcODB1Dks9%2B7TGzuJ6UHe5FtZjXxSPYSQQA%3D%3D%3D))